crates/ruff_linter/src/rules/ruff/rules/invalid_rule_code.rs

```diff
@@ -0,0 +1,159 @@
+use crate::noqa::{Code, Directive};
+use crate::registry::Rule;
+use crate::Locator;
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_text_size::Ranged;
+
+use crate::noqa::{Codes, NoqaDirectiveLine, NoqaDirectives};
+
+/// ## What it does
+/// Checks for `noqa` codes that are invalid.
+///
+/// ## Why is this bad?
+/// Invalid rule codes serve no purpose and may indicate outdated code suppressions.
+///
+/// ## Example
+/// ```python
+/// import os  # noqa: XYZ999
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import os
+/// ```
+///
+/// Or if there are still valid codes needed:
+/// ```python
+/// import os  # noqa: E402
+/// ```
+
+#[derive(ViolationMetadata)]
+pub(crate) struct InvalidRuleCode {
+    pub(crate) rule_code: String,
+}
+
+impl AlwaysFixableViolation for InvalidRuleCode {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Invalid rule code in `# noqa`: {}", self.rule_code)
+    }
+
+    fn fix_title(&self) -> String {
+        "Remove the rule code".to_string()
+    }
+}
+
+/// RUF102 for invalid noqa codes
+pub(crate) fn invalid_noqa_code(
+    diagnostics: &mut Vec<Diagnostic>,
+    noqa_directives: &NoqaDirectives,
+    locator: &Locator,
+    external: &[String],
+) {
+    for line in noqa_directives.lines() {
+        let Directive::Codes(directive) = &line.directive else {
+            continue;
+        };
+
+        let mut invalid_code_refs = vec![];
+        let mut all_invalid = true;
+
+        for code in directive.iter() {
+            let code_str = code.as_str();
+            if external.iter().any(|ext| code_str.starts_with(ext))
+                || Rule::from_code(code.as_str()).is_ok()
```

Comment:
```suggestion
            if Rule::from_code(code.as_str()).is_ok() || external.iter().any(|ext| code_str.starts_with(ext))
```

This might be faster because it checks the common case first.